### PR TITLE
Added bit twiddling hacks proofs using bv_decide

### PR DESCRIPTION
**Pull request by nwetzler**
### Description:

Evaluation of `bv_decide` on [Bit Twiddling Hacks](https://graphics.stanford.edu/~seander/bithacks.html).

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine?

make_all

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Proofs/Bit_twiddling.lean [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+-- (https://graphics.stanford.edu/~seander/bithacks.html#CopyIntegerSign)
+
+-- int v;      // we want to find the sign of v
+-- int sign;   // the result goes here
+
+-- // CHAR_BIT is the number of bits per byte (normally 8).
+-- sign = -(v < 0);  // if v < 0 then -1, else 0.
+-- // or, for one less instruction (but not portable):
+-- sign = v >> (sizeof(int) * CHAR_BIT - 1);
+
```
Comment (bollu) [inline Proofs/Bit_twiddling.lean:41]:
```suggestion
/--
1. Compute the sign of an integer
(https://graphics.stanford.edu/~seander/bithacks.html#CopyIntegerSign)

int v;      // we want to find the sign of v
int sign;   // the result goes here

// CHAR_BIT is the number of bits per byte (normally 8).
sign = -(v < 0);  // if v < 0 then -1, else 0.
// or, for one less instruction (but not portable):
sign = v >> (sizeof(int) * CHAR_BIT - 1);
-/
```

File: Proofs/Bit_twiddling.lean [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+-- ============================================================
+
+-- 2. Detect if two integers have opposite signs
+-- (https://graphics.stanford.edu/~seander/bithacks.html#DetectOppositeSigns)
+
+-- int x, y;               // input values to compare signs
+
+-- bool f = ((x ^ y) < 0); // true iff x and y have opposite signs
+
```
Comment (bollu) [inline Proofs/Bit_twiddling.lean:62]:
```suggestion
/--
2. Detect if two integers have opposite signs
(https://graphics.stanford.edu/~seander/bithacks.html#DetectOppositeSigns)

int x, y;               // input values to compare signs

bool f = ((x ^ y) < 0); // true iff x and y have opposite signs
-/
```

File: Proofs/Bit_twiddling.lean [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+-- ============================================================
+
+-- 3. Compute the integer absolute value (abs) without branching
+-- (https://graphics.stanford.edu/~seander/bithacks.html#IntegerAbs)
+
+-- int v;           // we want to find the absolute value of v
+-- unsigned int r;  // the result goes here
+-- int const mask = v >> sizeof(int) * CHAR_BIT - 1;
+
+-- r = (v + mask) ^ mask;
```
[the remark this answers is not in the document]
Comment (bollu) [reply Proofs/Bit_twiddling.lean:76]: And similarly for now and all of them.

**Review by bollu — COMMENTED** [review]
LGTM! Modulo the comment syntax :)

File: Proofs/Bit_twiddling.lean [hunk 4]
```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:2]:
```suggestion
Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: Proofs/Bit_twiddling.lean [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Nathan Wetzler
+-/
+
+import LeanSAT
+
+section bit_twiddling
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:9]:
```suggestion
namespace BitTwiddlingTests
```

File: Proofs/Bit_twiddling.lean [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+  let x4 := x3 &&& 0x0000000f#32
+  (0x00006996#32 >>> x4) &&& 1#32
+
+theorem parity32_correct (x : BitVec 32) :
+  (parity32_spec x) = ((parity32_impl x).getLsb 0) := by
+  unfold parity32_spec parity32_impl
+  repeat (unfold parity32_spec_rec)
+  bv_decide
+
+end bit_twiddling
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:259]:
```suggestion
end BitTwiddlingTests
```

File: Proofs/Bit_twiddling.lean [hunk 7]
```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Nathan Wetzler
+-/
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:5]:
```suggestion

The goal is to stress-test LeanSAT's performance using some classic 
bit-twiddling problems as benchmarks.
-/
```

File: Proofs/Bit_twiddling.lean [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+-- int x;  // we want to find the minimum of x and y
+-- int y;
+-- int r;  // the result goes here
+
+-- r = y ^ ((x ^ y) & -(x < y)); // min(x, y)
+
+-- To find the maximum, use:
+
+-- r = x ^ ((x ^ y) & -(x < y)); // max(x, y)
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:101]:
```suggestion
/--
4. Compute the minimum (min) or maximum (max) of two integers without branching
(https://graphics.stanford.edu/~seander/bithacks.html#IntegerMinOrMax)

int x;  // we want to find the minimum of x and y
int y;
int r;  // the result goes here

r = y ^ ((x ^ y) & -(x < y)); // min(x, y)

To find the maximum, use:

r = x ^ ((x ^ y) & -(x < y)); // max(x, y)
-/
```

File: Proofs/Bit_twiddling.lean [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+-- (https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2)
+
+-- unsigned int v; // we want to see if v is a power of 2
+-- bool f;         // the result goes here
+
+-- f = (v & (v - 1)) == 0;
+
+-- Note that 0 is incorrectly considered a power of 2 here. To remedy this, use:
+
+-- f = v && !(v & (v - 1));
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:131]:
```suggestion
/--
5. Determining if an integer is a power of 2
(https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2)

unsigned int v; // we want to see if v is a power of 2
bool f;         // the result goes here

f = (v & (v - 1)) == 0;

Note that 0 is incorrectly considered a power of 2 here. To remedy this, use:

f = v && !(v & (v - 1));
-/
```

File: Proofs/Bit_twiddling.lean [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+-- ============================================================
+
+-- 6. Conditionally set or clear bits without branching
+-- (https://graphics.stanford.edu/~seander/bithacks.html#ConditionalSetOrClearBitsWithoutBranching)
+
+-- bool f;         // conditional flag
+-- unsigned int m; // the bit mask
+-- unsigned int w; // the word to modify:  if (f) w |= m; else w &= ~m;
+
+-- w ^= (-f ^ w) & m;
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:152]:
```suggestion
/--
6. Conditionally set or clear bits without branching
(https://graphics.stanford.edu/~seander/bithacks.html#ConditionalSetOrClearBitsWithoutBranching)

bool f;         // conditional flag
unsigned int m; // the bit mask
unsigned int w; // the word to modify:  if (f) w |= m; else w &= ~m;

w ^= (-f ^ w) & m;
-/
```

File: Proofs/Bit_twiddling.lean [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+-- ============================================================
+
+-- 7. Conditionally negate a value without branching
+-- (https://graphics.stanford.edu/~seander/bithacks.html#ConditionalNegate)
+
+-- bool fNegate;  // Flag indicating if we should negate v.
+-- int v;         // Input value to negate if fNegate is true.
+-- int r;         // result = fNegate ? -v : v;
+
+-- r = (v ^ -fNegate) + fNegate;
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:168]:
```suggestion
/--
7. Conditionally negate a value without branching
(https://graphics.stanford.edu/~seander/bithacks.html#ConditionalNegate)

bool fNegate;  // Flag indicating if we should negate v.
int v;         // Input value to negate if fNegate is true.
int r;         // result = fNegate ? -v : v;

r = (v ^ -fNegate) + fNegate;
-/
```

File: Proofs/Bit_twiddling.lean [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+-- 8. Merge bits from two values according to a mask
+-- (https://graphics.stanford.edu/~seander/bithacks.html#MaskedMerge)
+
+-- unsigned int a;    // value to merge in non-masked bits
+-- unsigned int b;    // value to merge in masked bits
+-- unsigned int mask; // 1 where bits from b should be selected; 0 where from a.
+-- unsigned int r;    // result of (a & ~mask) | (b & mask) goes here
+
+-- r = a ^ ((a ^ b) & mask);
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:185]:
```suggestion
/--
8. Merge bits from two values according to a mask
(https://graphics.stanford.edu/~seander/bithacks.html#MaskedMerge)

unsigned int a;    // value to merge in non-masked bits
unsigned int b;    // value to merge in masked bits
unsigned int mask; // 1 where bits from b should be selected; 0 where from a.
unsigned int r;    // result of (a & ~mask) | (b & mask) goes here

r = a ^ ((a ^ b) & mask);
-/
```

File: Proofs/Bit_twiddling.lean [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+  bv_decide
+
+-- ============================================================
+
+-- 9. Counting bits set, in parallel
+-- (https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel)
+
+-- v = v - ((v >> 1) & 0x55555555);                    // reuse input as temporary
+-- v = (v & 0x33333333) + ((v >> 2) & 0x33333333);     // temp
+-- c = ((v + (v >> 4) & 0xF0F0F0F) * 0x1010101) >> 24; // count
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:198]:
```suggestion
/--
9. Counting bits set, in parallel
(https://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel)

v = v - ((v >> 1) & 0x55555555);                    // reuse input as temporary
v = (v & 0x33333333) + ((v >> 2) & 0x33333333);     // temp
c = ((v + (v >> 4) & 0xF0F0F0F) * 0x1010101) >> 24; // count
-/
```

File: Proofs/Bit_twiddling.lean [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+-- 10. Compute parity in parallel
+-- (https://graphics.stanford.edu/~seander/bithacks.html#ParityParallel)
+
+-- unsigned int v;  // word value to compute the parity of
+-- v ^= v >> 16;
+-- v ^= v >> 8;
+-- v ^= v >> 4;
+-- v &= 0xf;
+-- return (0x6996 >> v) & 1;
```
Comment (shigoel) [inline Proofs/Bit_twiddling.lean:233]:
```suggestion
/--
10. Compute parity in parallel
(https://graphics.stanford.edu/~seander/bithacks.html#ParityParallel)

unsigned int v;  // word value to compute the parity of
v ^= v >> 16;
v ^= v >> 8;
v ^= v >> 4;
v &= 0xf;
return (0x6996 >> v) & 1;
-/
```

**Review by shigoel — APPROVED** [review]

**Pull request merged**